hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -1658,16 +1676,19 @@ private ByteBuff allocate(int size, boolean intoHeap) {
     protected HFileBlock readBlockDataInternal(FSDataInputStream is, long offset,
       long onDiskSizeWithHeaderL, boolean pread, boolean verifyChecksum, boolean updateMetrics,
       boolean intoHeap) throws IOException {
-      if (offset < 0) {
-        throw new IOException("Invalid offset=" + offset + " trying to read " + "block (onDiskSize="
-          + onDiskSizeWithHeaderL + ")");
-      }
-
       final Span span = Span.current();
       final AttributesBuilder attributesBuilder = Attributes.builder();
       Optional.of(Context.current()).map(val -> val.get(CONTEXT_KEY))
         .ifPresent(c -> c.accept(attributesBuilder));
-      int onDiskSizeWithHeader = checkAndGetSizeAsInt(onDiskSizeWithHeaderL, hdrSize);
+      if (offset < 0) {
+        throw new IOException("Invalid offset=" + offset + " trying to read " + "block (onDiskSize="
+          + onDiskSizeWithHeaderL + ")");
+      }
+      if (!checkCallerProvidedOnDiskSizeWithHeader(onDiskSizeWithHeaderL)) {
+        throw new IOException(
```

Comment:
this is just an idea, what do you think -- what if, if the passed in size is invalid... we just ignore it? I think it's just an optimization, and below we can fetch the real size and handle any checksum issues (the most likely culprit for this being wrong).  I'm just thinking this might be another inevitable patch if one of those 2 risky sources of onDiskSizeWithHeaderL ever get loaded with a corrupt value.